### DISCUSSION TITLE
Bundling content questions.

### DISCUSSION THREAD

**Discussion by Blady-Com** [jralls/gtk-osx-build#73] [Q&A] [answered]
Hello,

This post may be off topic here, if so please tell me where could I post :-)

Regarding bundling [instructions](https://github.com/jralls/gtk-mac-bundler/blob/master/README.md) and [example](https://github.com/jralls/gtk-mac-bundler/blob/master/examples/gtk3-demo.bundle), here are some questions:

1. Folders `$(prefix)/lib/girepository-1.0 `and `${prefix}/share/gir-1.0` are not mention, shouldn't be ?
2. Shouldn't `loaders.cache` and `immodules.cache` be copied to the bundle?
3. Shouldn't `$(prefix)/bin/gdk-pixbuf-query-loaders` be copied to the bundle?
4. Shouldn't `PkgInfo` is not mention, shouldn't be ?
5. Shouldn't `$(prefix)/etc/fonts` be copied to the bundle?

Thanks, Pascal.

**Comment by jralls** [comment, marked as answer]
There wasn't really enough volume on gtk-osx-users to justify setting up discussions on all three repos, so this is The One. Given the nature of what you're asking about you could also have [raised an issue](https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues). Remember that these repos are just mirrors of the official ones on Gnome's Gitlab.

But you're here, so let's press on.

More complete and up-to-date instructions are maintained on the [Gnome wiki](https://wiki.gnome.org/Projects/GTK/OSX/Bundling). Thanks to markdown and web-based repo viewers it has become popular to use Readme files in the repo as primary documentation, but the wiki already exists and it's less overhead to edit it than the Readme file, so the wiki is the primary document.

Both are required for python, neither are required for C projects (and the example you pointed to is a C project; there's [another example](https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/blob/master/examples/python-launcher.bundle) for Python).

No, the caches created during build point to the build directory. The bundler runs the relevant query programs to build new caches with the bundle paths.

No, once signed one can't add binaries to the bundle so there's no reason for the bundle's user to regenerate loaders.cache.

PkgInfo is for PyPI. What has it got to do with macOS application bundles?

If your project allows users to change fonts then yes, you probably want to install the fontconfig setup files. That's not true of all projects.


loaders.cache and immodules.cache are regenerated by the bundler.

**Comment by Blady-Com** [comment]
Thanks for these answers.
> If your project allows users to change fonts then yes, you probably want to install the fontconfig setup files. That's not true of all projects.

Yes it does, is it correct to add only these lines?
```
  <data>
    ${prefix}/etc/fonts
  </data>
```

Another question about "dest" attribute, I saw this [usage](https://github.com/jralls/gtk-mac-bundler/blob/master/examples/gtk3-demo.bundle#L115) where it points a folder:
`  <data dest="${bundle}/Contents/Resources">`

But I failed with:
```
  <binary dest="${bundle}/Contents/Resources/bin">
    ${prefix}/bin/ada_language_server
  </binary>
```
The bundler has created a file "bin".
What is correct?

**Reply by jralls** [reply]
Both are correct. Just like with the `mv` or `cp` commands, if the source is a single file and the destination is an already-created directory then the file is copied into that directory; if the destination is an existing file it is replaced, and if it doesn't exist it is created as a file. If the source is multiple files (e.g. `${prefix}/bin/ada*` where there is more than one file in `${prefix}/bin/` whose name begins with `ada` then `cp` will create a directory and copy all of the files into it. I *think* that works here too. 

`rsync` has some nice rules where a trailing `/` is significant in determining whether a directory is created and how multiple files are copied. It would be cool to have something similar here, but I don't have time to implement it. PR welcome. ;-)

**Reply by jralls** [reply]
> Yes it does, is it correct to add only these lines?

It is as long as you've built fontconfig with gtk-osx so that `etc/fonts` has the right contents.

**Comment by Blady-Com** [comment]
Thanks John for your answers in this post and for all your previous answers.
I've been able to port a huge IDE for Ada language [GNAT Studio](https://github.com/AdaCore/gnatstudio) and create a [macOS app](https://sourceforge.net/projects/gnuada/files/GNAT_GPL%20Mac%20OS%20X/2022-monterey) :-)
Now I have to take a step back and to gather all my notes.
I will go further with bundling issues on [gtk-mac-bundler](https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues).
I'll surely still need your help ;-).
Thanks again, Pascal.
